module/VuFind/src/VuFind/Export/Zotero/ZoteroService.php

```diff
@@ -0,0 +1,395 @@
+                    ]
+                )
+            );
+        $this->accessTokenService->persistEntity($accessToken);
+        // Perform the actual export:
+        return $this->exportToZotero($accessToken);
+    }
+
+    /**
+     * Get temporary credentials from Zotero and return an authorization URL.
+     *
+     * @return string
+     *
+     * @throws ZoteroException
+     */
+    protected function getZoteroAuthUrlWithTempCredentials(): string
+    {
+        // Get request token from Zotero:
+        try {
+            $this->sessionContainer->tempCredentials = $this->zoteroOAuth->getTemporaryCredentials();
+        } catch (CredentialsException $e) {
+            $this->logError('Could not retrieve temporary credentials from Zotero: ' . (string)$e);
+            throw new ZoteroException('Could not retrieve temporary credentials from Zotero');
+        }
+
+        // Redirect to Zotero for authentication:
+        return $this->zoteroOAuth->getAuthorizationUrl($this->sessionContainer->tempCredentials);
+    }
+
+    /**
+     * Export record(s) to Zotero using its v3 API.
+     *
+     * @param AccessTokenEntityInterface $accessToken Access token with Zotero credentials
+     *
+     * @return int Exported record count
+     *
+     * @throws ZoteroException
+     */
+    protected function exportToZotero(AccessTokenEntityInterface $accessToken): int
+    {
+        if (!($callbackUrl = $this->sessionContainer->exportCallbackUrl)) {
+            $this->logError('Export callback URL not available in session');
+            throw new ZoteroException('Export callback URL not available in session');
+        }
+
+        try {
+            $vufindResponse = $this->httpService->createClient($callbackUrl)->request('GET', $callbackUrl);
+        } catch (\Exception $e) {
+            $this->logError("GET request for '$callbackUrl' failed: " . (string)$e);
+            throw new ZoteroException('Export request failed');
+        }
+
+        if ($vufindResponse->getStatusCode() !== 200) {
+            $this->logError(
+                "GET request for '$callbackUrl' failed: "
+                . $vufindResponse->getStatusCode() . ': ' . $vufindResponse-> getReasonPhrase()
+            );
+            throw new ZoteroException('Export request failed');
+        }
+
+        if (!($records = json_decode($vufindResponse->getBody()->getContents()))) {
+            $this->logError('Could not decode VuFind export response as JSON: ' . $vufindResponse->getBody());
+            throw new ZoteroException('Could not decode export response');
+        }
+
+        $credentials = json_decode($accessToken->getData(), true);
+        $zoteroUserId = $credentials['zoteroUserId'];
+        $zoteroApiKey = $credentials['zoteroApiKey'];
+
+        $apiUrl = "https://api.zotero.org/users/$zoteroUserId/items";
+        // Make sure we send an array of records for a single one as well:
+        $records = is_array($records) ? $records : [$records];
```

Comment:
Will this work, or are there weird types involved that make this problematic?


```suggestion
        $records = (array)$records;
```

Reply:
That won't work. Casting an object into an array will create an array from the object's public properties.

Reply:
I had a feeling that was the answer, but my brain just always wants to optimize that statement. It's annoying! ;-)